### PR TITLE
new programmatic-reference page

### PR DESCRIPTION
**Pull request by sarah11918**
#### Description (required)

Moves advanced API section from CLI ref to its own page

- Advanced API section moved from cli-reference.mdx
- Existing text copy/paste into new file as-is **except for heading levels** which have all changed by one and **type formatting** (Thank you, Armand!)
- puts this page in sidebar. (Will also need to be added to #9890 in the "Other API" section.)

Does NOT 
- contain a redirect, because we can't redirect *from* anchor slugs.

### CODE REVIEW

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+```ts
+export interface DevServer {
+	address: AddressInfo;
+	handle: (req: http.IncomingMessage, res: http.ServerResponse<http.IncomingMessage>) => void;
+	watcher: vite.FSWatcher;
+	stop(): Promise<void>;
+}
+```
+
+#### `address`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:100]:
```suggestion
#### `address`

<p>

**Type:** `AddressInfo`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+The address the dev server is listening on.
+
+This property contains the value returned by Node's [`net.Server#address()` method](https://nodejs.org/api/net.html#serveraddress).
+
+#### `handle()`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:106]:
```suggestion
#### `handle()`

<p>

**Type:** `(req: http.IncomingMessage, res: http.ServerResponse<http.IncomingMessage>) => void`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+A handle for raw Node HTTP requests. You can call `handle()` with an [`http.IncomingMessage`](https://nodejs.org/api/http.html#class-httpincomingmessage) and an [`http.ServerResponse`](https://nodejs.org/api/http.html#class-httpserverresponse) instead of sending a request through the network.
+
+#### `watcher`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:110]:
```suggestion
#### `watcher`

<p>

**Type:** `vite.FSWatcher`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+The [Chokidar file watcher](https://github.com/paulmillr/chokidar#getting-started) as exposed by [Vite's development server](https://vite.dev/guide/api-javascript#vitedevserver).
+
+#### `stop()`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:114]: I have a doubt here. Usually for functions we use the following format: `() => Promise<void>` but the interface above does not match. I checked in [`withastro/astro` repo](https://github.com/withastro/astro/blob/25baa4ed0c5f55fa85c2c7e2c15848937ed1dc9b/packages/astro/src/core/dev/dev.ts#L24) and... TIL, `stop(): Promise<void>` seems valid. So I don't know if we should use:
* `**Type:** () => Promise<void>`
* or `**Type:** Promise<void>`

```suggestion
#### `stop()`

<p>

**Type:** `Promise<void>`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+```ts
+export interface PreviewServer {
+	host?: string;
+	port: number;
+	closed(): Promise<void>;
+	stop(): Promise<void>;
+}
+```
+
+#### `host`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:165]:
```suggestion
#### `host`

<p>

**Type:** `string`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+The host where the server is listening for connections.
+
+Adapters are allowed to leave this field unset. The value of `host` is implementation-specific.
+
+#### `port`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:171]:
```suggestion
#### `port`

<p>

**Type:** `number`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+The port where the server is listening for connections.
+
+#### `stop()`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:175]: Same remark as before regarding the format to use... `() => ...` or directly the output.
```suggestion
#### `stop()`

<p>

**Type:** `Promise<void>`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+Asks the preview server to close, stop accepting requests, and drop idle connections.
+
+The returned `Promise` resolves when the close request has been sent. This does not mean that the server has closed yet. Use the [`closed()`](#closed) method if you need to ensure the server has fully closed.
+
+#### `closed()`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:181]: Same remark as before regarding the format to use... `() => ...` or directly the output.
```suggestion
#### `closed()`

<p>

**Type:** `Promise<void>`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+- `"debug"`: Log everything, including noisy debugging diagnostics.
+- `"info"`: Log informational messages, warnings, and errors.
+- `"warn"`: Log warnings and errors.
+- `"error"`: Log errors only.
+- `"silent"`: No logging.
+
+## `dev()`
+
+**Type:** `(inlineConfig: AstroInlineConfig) => Promise<DevServer>`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:74]: If we add the other types, nit to be consistent:
```suggestion
<p>

**Type:** `(inlineConfig: AstroInlineConfig) => Promise<DevServer>`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+#### `stop()`
+
+Stops the development server. This closes all idle connections and stops listening for new connections.
+
+Returns a `Promise` that resolves once all pending requests have been fulfilled and all idle connections have been closed.
+
+## `build()`
+
+**Type:** `(inlineConfig: AstroInlineConfig) => Promise<void>`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:122]: If we add the other types, nit to be consistent:
```suggestion
<p>

**Type:** `(inlineConfig: AstroInlineConfig) => Promise<void>`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+import { build } from "astro";
+
+await build({
+  root: "./my-project",
+});
+```
+
+## `preview()`
+
+**Type:** `(inlineConfig: AstroInlineConfig) => Promise<PreviewServer>`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:136]: If we add the other types, nit to be consistent:

```suggestion
<p>

**Type:** `(inlineConfig: AstroInlineConfig) => Promise<PreviewServer>`
</p>
```

File: src/content/docs/en/reference/programmatic-reference.mdx [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+The returned `Promise` resolves when the close request has been sent. This does not mean that the server has closed yet. Use the [`closed()`](#closed) method if you need to ensure the server has fully closed.
+
+#### `closed()`
+
+Returns a `Promise` that will resolve once the server is closed and reject if an error happens on the server.
+
+## `sync()`
+
+**Type:** `(inlineConfig: AstroInlineConfig) => Promise<void>`
```
Comment (ArmandPhilippot) [inline src/content/docs/en/reference/programmatic-reference.mdx:187]: If we add the other types, nit to be consistent:
```suggestion
<p>

**Type:** `(inlineConfig: AstroInlineConfig) => Promise<void>`
</p>
```

**Review by ArmandPhilippot — COMMENTED** [review]
LGTM! But with these changes, the information on the new page are no longer accessible until the new sidebar...

> Temporary workaround could be to put this in the current sidebar now, or to live with it not in the sidebar temporarily

So, yes I think putting it in the current sidebar in the meantime could be a good thing and `Other Development APIs` seems to me to be a good candidate.

I also suggested some `Type:` on subheadings. The CLI page is not updated with the API reference format, but since we create a new page, I think we could take advantage of this to add those information here.

I have not added any suggestions for `AstroInlineConfig` and `DevServer` because we already present the type as code, it seemed redundant to me... and perhaps more readable in these cases.

File: src/content/docs/en/reference/programmatic-reference.mdx (hunk 4 above)
Comment (sarah11918) [reply src/content/docs/en/reference/programmatic-reference.mdx:114]: Alright, then I say we use this until someone tells us it's wrong! 😅

File: src/content/docs/en/reference/programmatic-reference.mdx (hunk 7 above)
Comment (sarah11918) [reply src/content/docs/en/reference/programmatic-reference.mdx:175]: Right, let's keep these the same.

**Comment by sarah11918** [comment]
Thank you for your thorough review @ArmandPhilippot ! I think your default choices made sense, and we can always change them if necessary!

**Pull request merged**